source/server/drain_manager_impl.cc

```diff
@@ -17,25 +17,27 @@ namespace Server {
 DrainManagerImpl::DrainManagerImpl(Instance& server,
                                    envoy::config::listener::v3::Listener::DrainType drain_type,
                                    Event::Dispatcher& dispatcher)
-    : server_(server), dispatcher_(dispatcher), drain_type_(drain_type), children_(dispatcher) {}
+    : server_(server), dispatcher_(dispatcher), drain_type_(drain_type),
+      children_(Common::ThreadSafeCallbackManager::create()) {}
 
-DrainManagerSharedPtr
+DrainManagerPtr
 DrainManagerImpl::createChildManager(Event::Dispatcher& dispatcher,
                                      envoy::config::listener::v3::Listener::DrainType drain_type) {
-  auto child = std::make_shared<DrainManagerImpl>(server_, drain_type, dispatcher);
-
-  // Wire up the child so that when the parent starts draining, the child also sees the state-change
-  auto child_cb = children_.add(dispatcher, [child_ = std::weak_ptr<DrainManagerImpl>(child)] {
-    auto child = child_.lock();
-    if (child && !child->draining_) {
-      child->startDrainSequence([] {});
-    }
-  });
-  child->parent_callback_handle_ = child_cb;
+  auto child = std::make_unique<DrainManagerImpl>(server_, drain_type, dispatcher);
+
+  // Wire up the child so that when the parent starts draining, the child also sees the
+  // state-change
+  auto child_cb = children_->add(
+      dispatcher, [child = child.get(), still_alive = std::weak_ptr<bool>(child->still_alive_)] {
```

Comment:
This still_alive shouldn't be needed now that we have a still_alive at the CallbackHolder level that protects against execution of callbacks that have been unregistered.

Reply:
Ah, good call!